In [1]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn import datasets

$ Sigmoid  = 1/{(1 + e^{-z})} $

In [2]:
def sigmoid(predictions):
    return 1/(1 + np.exp(-predictions))
    

### Logistic Regression Class 

In [ ]:
class LogisticRegression:
    
    def __init__(self, lr=0.001, num_iters=1000):
        self.lr = lr
        self.num_iters = num_iters
        
        self.weight = None
        self.bias = None
        
    def fit(self, X,y):
        num_samples, num_features = X.shape
        self.weight = np.zeros(num_features)
        self.bias = 0
        
        for _ in range(self.num_iters):
            linear_predictions = np.dot(X , self.weight) + self.bias
            logistic_predictions = sigmoid(linear_predictions)
            
            #Compute the gradients
            dw = (1 / num_samples) * np.dot(X.T , (logistic_predictions - y))
            db = (1/ num_samples) * np.sum(logistic_predictions - y)
            
            #Update the parameters
            self.weight = self.weight - self.lr * dw
            self.bias = self.bias - self.lr * db
            
    def predict(self, X):
        linear_prediction = np.dot(X , self.weight) + self.bias
        logistic_prediction = sigmoid(linear_prediction)
        class_prediction = [0 if y<=0.5 else 1 for y in logistic_prediction]
        return class_prediction
          

## Model Inference

#### Loading the dataset

In [4]:
dataset = datasets.load_breast_cancer()
X, y = dataset.data , dataset.target
X_train,X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 1234)

#### Initializing our Logistic Regression Class

In [5]:
#Function to calculate the accuracy of our Algorithm
def accuracy(y_pred, label):
    return np.sum(y_pred == label)/ len(label)

#### Fitting our data in the Model

In [8]:
log_reg = LogisticRegression()
log_reg.fit(X_train, y_train)
predictions = log_reg.predict(X_test) 

acc = accuracy(predictions, y_test)
print("="*25, f"Accuracy:- {acc*100:.5f}%", "="*25)

C:\Users\athar\AppData\Local\Temp\ipykernel_5784\2834829709.py:2: RuntimeWarning: overflow encountered in exp
  return 1/(1 + np.exp(-predictions))


========================= Accuracy:- 89.47368% =========================


- We are getting 89% accuracy, We can try to change the learning rate to see if we get better accuracy or not

#### Hyperparameter Tuning

In [9]:
log_reg_2 = LogisticRegression(lr=0.01)
log_reg_2.fit(X_train, y_train)
predictions2 = log_reg_2.predict(X_test)

acc2 = accuracy(predictions2, y_test)
print("="*25, f"Accuracy= {acc2*100:.5f}%", "="*25)
    

C:\Users\athar\AppData\Local\Temp\ipykernel_5784\2834829709.py:2: RuntimeWarning: overflow encountered in exp
  return 1/(1 + np.exp(-predictions))


========================= Accuracy= 92.10526% =========================


- We got 92% accuracy which is good for our model and hence we can conclude that our Implementation is working correctly.